dashboard/app/controllers/aidiff_threads_controller.rb

```diff
@@ -244,6 +256,44 @@ def chat_completion
     end
   end
 
+  # Returns name(s) of programming environment(s) in the chat context
+  private def get_labs(context_type)
+    if context_type == SharedConstants::AI_DIFF_CONTEXT[:LEVEL]
+      labs_from_level(@level)
+    elsif context_type == SharedConstants::AI_DIFF_CONTEXT[:LESSON]
+      @lesson.levels.map {|level| labs_from_level(level)}
+    elsif context_type == SharedConstants::AI_DIFF_CONTEXT[:UNIT]
+      @unit.levels.map {|level| labs_from_level(level)}
+    elsif context_type == SharedConstants::AI_DIFF_CONTEXT[:COURSE]
+      @unit_group.default_units.map do |unit|
+        unit.levels.map {|level| labs_from_level(level)}
```

Comment:
```suggestion
        get_labs_for_unit(unit)
```
Optional nit: this logic is duplicated 3 times.  We could refactor into a helper function to help make it clear what is going on.